### ISSUE TITLE
Vim9: not easy to find where whitespace is missing in dictionary

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7792]
**Describe the bug**

In Vim9 script, it is sometimes not easy to find where a whitespace is missing inside a dictionary.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9
        echo {a: 1, b:2, c: 3}
    EOF
    )

`E1069` is raised with this message:

    E1069: White space required after ':'

This is expected; but no context is given about where the whitespace is missing.

**Expected behavior**

The error message gives us more context about where the whitespace is missing.

**Environment**

 - Vim version: 8.2 Included patches: 1-2475
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(365)

**Additional context**

Here is another similar snippet where the error message does give us more context:
```vim
vim9
echo {a: 1 ? 'b': 'c'}
```
    E1004: White space required before and after ':' at ": 'c'}"
                                                 ^-------------^
                                                        ✔

This is inconsistent.  If Vim gives us more context in this last snippet, then it should do the same in the first one.

---

A similar issue was fixed in [8.2.2266](https://github.com/vim/vim/releases/tag/v8.2.2266). I tried to copy the relevant code, and came up with this patch:
```diff
diff --git a/src/dict.c b/src/dict.c
index 260f22985..516a27d0e 100644
--- a/src/dict.c
+++ b/src/dict.c
@@ -970,7 +970,7 @@ eval_dict(char_u **arg, typval_T *rettv, evalarg_T *evalarg, int literal)
 	}
 	if (vim9script && (*arg)[1] != NUL && !VIM_ISWHITE((*arg)[1]))
 	{
-	    semsg(_(e_white_space_required_after_str), ":");
+	    semsg(_(e_white_space_required_after_str_at_str), ":", *arg);
 	    clear_tv(&tvkey);
 	    goto failret;
 	}
@@ -1012,7 +1012,7 @@ eval_dict(char_u **arg, typval_T *rettv, evalarg_T *evalarg, int literal)
 	{
 	    if (vim9script && (*arg)[1] != NUL && !VIM_ISWHITE((*arg)[1]))
 	    {
-		semsg(_(e_white_space_required_after_str), ",");
+		semsg(_(e_white_space_required_after_str_at_str), ",", *arg);
 		goto failret;
 	    }
 	    *arg = skipwhite(*arg + 1);
diff --git a/src/errors.h b/src/errors.h
index e6d7ab26c..6a11afa56 100644
--- a/src/errors.h
+++ b/src/errors.h
@@ -175,8 +175,8 @@ EXTERN char e_separator_mismatch_str[]
 	INIT(= N_("E1067: Separator mismatch: %s"));
 EXTERN char e_no_white_space_allowed_before_str[]
 	INIT(= N_("E1068: No white space allowed before '%s'"));
-EXTERN char e_white_space_required_after_str[]
-	INIT(= N_("E1069: White space required after '%s'"));
+EXTERN char e_white_space_required_after_str_at_str[]
+	INIT(= N_("E1069: White space required after '%s' at \"%s\""));
 EXTERN char e_missing_from[]
 	INIT(= N_("E1070: Missing \"from\""));
 EXTERN char e_invalid_string_after_from[]
diff --git a/src/list.c b/src/list.c
index c51c99bc9..afe322439 100644
--- a/src/list.c
+++ b/src/list.c
@@ -1311,7 +1311,7 @@ eval_list(char_u **arg, typval_T *rettv, evalarg_T *evalarg, int do_error)
 	{
 	    if (vim9script && !IS_WHITE_OR_NUL((*arg)[1]) && (*arg)[1] != ']')
 	    {
-		semsg(_(e_white_space_required_after_str), ",");
+		semsg(_(e_white_space_required_after_str_at_str), ",", *arg);
 		goto failret;
 	    }
 	    *arg = skipwhite(*arg + 1);
diff --git a/src/userfunc.c b/src/userfunc.c
index bdf406426..38ac73d80 100644
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -126,7 +126,7 @@ one_function_arg(
 	    ++p;
 	    if (!skip && !VIM_ISWHITE(*p))
 	    {
-		semsg(_(e_white_space_required_after_str), ":");
+		semsg(_(e_white_space_required_after_str_at_str), ":", p);
 		return arg;
 	    }
 	    type = skipwhite(p);
@@ -297,7 +297,7 @@ get_function_args(
 		if (!skip && in_vim9script()
 				      && !IS_WHITE_OR_NUL(*p) && *p != endchar)
 		{
-		    semsg(_(e_white_space_required_after_str), ",");
+		    semsg(_(e_white_space_required_after_str_at_str), ",", p);
 		    goto err_ret;
 		}
 	    }
@@ -487,7 +487,7 @@ skip_arrow(
 	    if (white_error != NULL && !VIM_ISWHITE(s[1]))
 	    {
 		*white_error = TRUE;
-		semsg(_(e_white_space_required_after_str), ":");
+		semsg(_(e_white_space_required_after_str_at_str), ":", s);
 		return NULL;
 	    }
 	    s = skipwhite(s + 1);
@@ -884,7 +884,7 @@ get_func_tv(
 	    break;
 	if (vim9script && !IS_WHITE_OR_NUL(argp[1]))
 	{
-	    semsg(_(e_white_space_required_after_str), ",");
+	    semsg(_(e_white_space_required_after_str_at_str), ",", argp);
 	    ret = FAIL;
 	    break;
 	}
diff --git a/src/vim9compile.c b/src/vim9compile.c
index 6a1d8d9e7..840ea9993 100644
--- a/src/vim9compile.c
+++ b/src/vim9compile.c
@@ -2807,7 +2807,7 @@ compile_arguments(char_u **arg, cctx_T *cctx, int *argcount)
 	{
 	    ++p;
 	    if (*p != NUL && !VIM_ISWHITE(*p))
-		semsg(_(e_white_space_required_after_str), ",");
+		semsg(_(e_white_space_required_after_str_at_str), ",", p);
 	}
 	else
 	    must_end = TRUE;
@@ -3073,7 +3073,7 @@ compile_list(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	    ++p;
 	    if (*p != ']' && !IS_WHITE_OR_NUL(*p))
 	    {
-		semsg(_(e_white_space_required_after_str), ",");
+		semsg(_(e_white_space_required_after_str_at_str), ",", p);
 		return FAIL;
 	    }
 	}
@@ -3242,7 +3242,7 @@ compile_dict(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	whitep = *arg + 1;
 	if (!IS_WHITE_OR_NUL(*whitep))
 	{
-	    semsg(_(e_white_space_required_after_str), ":");
+	    semsg(_(e_white_space_required_after_str_at_str), ":", *arg);
 	    return FAIL;
 	}
 
@@ -3279,7 +3279,7 @@ compile_dict(char_u **arg, cctx_T *cctx, ppconst_T *ppconst)
 	whitep = *arg + 1;
 	if (!IS_WHITE_OR_NUL(*whitep))
 	{
-	    semsg(_(e_white_space_required_after_str), ",");
+	    semsg(_(e_white_space_required_after_str_at_str), ",", *arg);
 	    return FAIL;
 	}
 	*arg = skipwhite(*arg + 1);
@@ -5711,7 +5711,7 @@ compile_lhs(
 	    // parse optional type: "let var: type = expr"
 	    if (!VIM_ISWHITE(var_end[1]))
 	    {
-		semsg(_(e_white_space_required_after_str), ":");
+		semsg(_(e_white_space_required_after_str_at_str), ":", var_end);
 		return FAIL;
 	    }
 	    p = skipwhite(var_end + 1);
diff --git a/src/vim9script.c b/src/vim9script.c
index 010b4bfdb..4cf5b9e1d 100644
--- a/src/vim9script.c
+++ b/src/vim9script.c
@@ -629,7 +629,7 @@ vim9_declare_scriptvar(exarg_T *eap, char_u *arg)
     }
     if (!VIM_ISWHITE(p[1]))
     {
-	semsg(_(e_white_space_required_after_str), ":");
+	semsg(_(e_white_space_required_after_str_at_str), ":", p);
 	return arg + STRLEN(arg);
     }
     name = vim_strnsave(arg, p - arg);
diff --git a/src/vim9type.c b/src/vim9type.c
index 1374cee1b..750b190a2 100644
--- a/src/vim9type.c
+++ b/src/vim9type.c
@@ -788,7 +788,7 @@ parse_type(char_u **arg, garray_T *type_gap, int give_error)
 			    if (!VIM_ISWHITE(*p))
 			    {
 				if (give_error)
-				    semsg(_(e_white_space_required_after_str), ",");
+				    semsg(_(e_white_space_required_after_str_at_str), ",", p);
 				return NULL;
 			    }
 			}
@@ -815,7 +815,7 @@ parse_type(char_u **arg, garray_T *type_gap, int give_error)
 		    // parse return type
 		    ++*arg;
 		    if (!VIM_ISWHITE(**arg) && give_error)
-			semsg(_(e_white_space_required_after_str), ":");
+			semsg(_(e_white_space_required_after_str_at_str), ":", *arg);
 		    *arg = skipwhite(*arg);
 		    ret_type = parse_type(arg, type_gap, give_error);
 		    if (ret_type == NULL)
```
It seems to fix the issue. I ran the tests and they all passed (a few were skipped); which is weird, I would have expected a few tests to fail considering that the error messages are now sometimes different...

**Comment by lacygoill** [comment]
Here is another case where it's not easy to find a missing whitespace:
```vim
vim9
def GetExpansion(...l: any): string
    return 'bbb'
enddef
def Func()
    var abbr = 'aaa'
    exe 'inorea <silent> ' .. abbr
         .. ' <c-r>=<sid>ExpandAdj(' .. string(abbr)
         .. ', ' .. GetExpansion(abbr,'adj')->string() .. ')<cr>'
enddef
Func()
```
    E1069: White space required after ','

The missing whitespace is here:

    .. ', ' .. GetExpansion(abbr,'adj')->string() .. ')<cr>'
                                ^
                                ✘

Not easy to find.  Especially because there is another comma before.

    .. ', ' .. GetExpansion(abbr,'adj')->string() .. ')<cr>'
        ^
        irrelevant

The previous patch improves the error message in this case too.

**Comment by brammool** [comment]
was fixed with patch 8.2.2480
